metadata_layer.py

```diff
@@ -351,36 +351,188 @@ def __repr__(self):
+                if link.rights_uri:
+                    self.set_default_rights_uri(data_source=self.data_source, default_rights_uri=link.rights_uri)
+
+                # An open-access link or open-access rights implies a FormatData object.
+                open_access_link = (link.rel == Hyperlink.OPEN_ACCESS_DOWNLOAD
+                                    and link.href)
+                
+                open_access_rights_link = (link.media_type in Representation.BOOK_MEDIA_TYPES 
+                                           and link.href
+                                           and self.default_rights_uri in RightsStatus.OPEN_ACCESS)
+                
+                #if open_access_link or open_access_rights_link:
+                if open_access_link:
+                    self.formats.append(
+                        FormatData(
+                            content_type=link.media_type,
+                            drm_scheme=DeliveryMechanism.NO_DRM,
+                            link=link
+                        )
+                )
+
+
+    def __repr__(self):
+        description_string = '<CirculationData primary_identifier="%r" licenses_owned="%s" licenses_available="%s" licenses_reserved=%s' % (
+            self.primary_identifier, self.licenses_owned, self.licenses_available, self.licenses_reserved
+        )
+        description_string += 'patrons_in_hold_queue="%s" default_rights_uri="%s" first_appearance="%s" last_checked="%s" >' % (
+            self.patrons_in_hold_queue, self.default_rights_uri, self.first_appearance, self.last_checked
+        )
+        description_string += 'links="%r" formats="%r" data_source_obj="%r">' % (
+            self.links, self.formats, self.data_source_obj
+        )
+
+        return description_string
+    
+
+    def data_source(self, _db):
+        if not self.data_source_obj:
+            if self._data_source:
+                print "CirculationData.data_source(): self._data_source=%s" % self._data_source
+                obj = DataSource.lookup(_db, self._data_source)
+                if not obj:
+                    raise ValueError("Data source %s not found!" % self._data_source)
+                if not obj.offers_licenses:
+                    raise ValueError("Data source %s does not offer licenses and cannot be used as a LicenseData data source." % self._data_source)
+            else:
+                obj = None
+            self.data_source_obj = obj
+        return self.data_source_obj
+
+    def license_pool(self, _db):
+        """Find or create a LicensePool object for this CirculationData."""
+        if not self.primary_identifier:
+            raise ValueError(
+                "Cannot find license pool: CirculationData has no primary identifier."
+            )
+
+        license_pool = None
+        is_new = False
+
+        identifier_obj, ignore = self.primary_identifier.load(_db)
+        data_source = self.data_source(_db)
```

Comment:
Looking at these two lines right next to each other it occurs to me we might be able to save some ugly code by creating a DataSourceData object which has a load() method. In the constructors we would just ensure that whatever was passed in we end up with a DataSourceData object, and then when we want the actual DataSource we call load(). Just something to think about.